src/backend/distributed/commands/schema_based_sharding.c

```diff
@@ -108,6 +123,85 @@ ShouldCreateTenantSchemaTable(Oid relationId)
 }
 
 
+/*
+ * EnsureTableKindSupportedForTenantSchema ensures that given table's kind is
+ * supported by a tenant schema.
+ */
+static void
+EnsureTableKindSupportedForTenantSchema(Oid relationId)
+{
+	if (IsForeignTable(relationId))
+	{
+		ereport(ERROR, (errmsg("cannot create a foreign table in a distributed "
+							   "schema")));
+	}
+	else if (PartitionTable(relationId))
+	{
+		ErrorIfIllegalPartitioningInTenantSchema(PartitionParentOid(relationId),
+												 relationId);
+	}
+	else if (IsChildTable(relationId) || IsParentTable(relationId))
+	{
+		ereport(ERROR, (errmsg("tables in a distributed schema cannot inherit or "
+							   "be inherited")));
+	}
+}
+
+
+/*
+ * EnsureFKeysForTenantTable ensures that all referencing and referenced foreign
+ * keys are allowed for given table.
+ */
+static void
+EnsureFKeysForTenantTable(Oid relationId)
+{
+	Oid tenantSchemaId = get_rel_namespace(relationId);
+	int fKeyReferencingFlags = INCLUDE_REFERENCING_CONSTRAINTS | INCLUDE_ALL_TABLE_TYPES;
+	List *referencingForeignKeys = GetForeignKeyOids(relationId, fKeyReferencingFlags);
+	Oid foreignKeyId = InvalidOid;
+	foreach_oid(foreignKeyId, referencingForeignKeys)
+	{
+		Oid referencedTableId = GetReferencedTableId(foreignKeyId);
+		Oid referencedTableSchemaId = get_rel_namespace(referencedTableId);
+
+		/* We allow foreign keys to a table in the same schema */
+		if (tenantSchemaId == referencedTableSchemaId)
+		{
+			continue;
+		}
+
+		/*
+		 * Allow foreign keys to the other schema only if the referenced table is
+		 * a reference table.
+		 */
+		if (!IsCitusTable(referencedTableId) ||
+			!IsCitusTableType(referencedTableId, REFERENCE_TABLE))
+		{
+			ereport(ERROR, errmsg("foreign keys from distributed schemas can only "
+								  "point to reference tables in regular schemas"));
+		}
+	}
+
+	int fKeyReferencedFlags = INCLUDE_REFERENCED_CONSTRAINTS | INCLUDE_ALL_TABLE_TYPES;
+	List *referencedForeignKeys = GetForeignKeyOids(relationId, fKeyReferencedFlags);
+	foreach_oid(foreignKeyId, referencedForeignKeys)
+	{
+		Oid referencingTableId = GetReferencingTableId(foreignKeyId);
+		Oid referencingTableSchemaId = get_rel_namespace(referencingTableId);
+
+		/* We allow foreign keys from a table in the same schema */
+		if (tenantSchemaId == referencingTableSchemaId)
+		{
+			continue;
+		}
+
+		/* Not allow any foreign keys from the other schema */
+		ereport(ERROR, errmsg("cannot create foreign keys to tables in a distributed "
+							  "schema from another schema"));
+	}
```

Comment:
For both ereports, can we also report name of the fkey constraint, referencing table name and referenced table name in errdetail?